File: wrsvm_package/wrsvm/classifier.py
"""Scikit-learn compatible WRSVM classifier.

Supports four decomposition strategies: Crammer-Singer (``"cs"``),
SimMSVM (``"simmsvm"``), One-vs-One (``"ovo"``), and One-vs-Rest (``"ovr"``).
"""

from __future__ import annotations

import numpy as np
from sklearn.base import BaseEstimator, ClassifierMixin
from sklearn.utils.validation import check_is_fitted, check_X_y

from wrsvm.decomposition import OVOClassifier, OVRClassifier
from wrsvm.simmsvm import predict_simmsvm, solve_simmsvm
from wrsvm.solver import predict as _predict_cs
from wrsvm.solver import solve_crammer_singer


class WRSVMClassifier(BaseEstimator, ClassifierMixin):
    """Weighted Relaxed SVM for multiclass classification.

    Solves one of four decomposition strategies:

    - ``"cs"`` — Crammer-Singer direct formulation (N * K dual variables)
    - ``"simmsvm"`` — Simultaneous multiclass with one dual per sample (N vars)
    - ``"ovo"`` — One-vs-One pairwise decomposition
    - ``"ovr"`` — One-vs-Rest decomposition

    Parameters
    ----------
    strategy : {"cs", "simmsvm", "ovo", "ovr"}, default "cs"
    C : float, default 100.0
    kernel : {"rbf", "linear", "poly", "sigmoid", "laplacian"}, default "rbf"
    gamma : float, default 0.1
        Bandwidth for rbf / laplacian / sigmoid; feature scale for poly.
        Ignored by the linear kernel.
    degree : int, default 3
        Polynomial degree. Only used when ``kernel="poly"``.
    coef0 : float, default 0.0
        Independent term for ``kernel in {"poly", "sigmoid"}``.
    upsilon : float, default 0.2
    solver : str, default "CLARABEL"
        One of ``"CLARABEL"``, ``"SCS"``, ``"SCS_GPU"``, ``"GUROBI"``, or any
        other name recognized by CVXPY. Falls back to SCS on failure.
    kernel_backend : {"numpy", "torch"}, default "numpy"
    """

    def __init__(self, strategy: str = "cs",
                 C: float = 100.0, gamma: float = 0.1, upsilon: float = 0.2,
                 kernel: str = "rbf", degree: int = 3, coef0: float = 0.0,
                 solver: str = "CLARABEL", kernel_backend: str = "numpy"):
        self.strategy = strategy
        self.C = C
        self.gamma = gamma
        self.upsilon = upsilon
        self.kernel = kernel
        self.degree = degree
        self.coef0 = coef0
        self.solver = solver
        self.kernel_backend = kernel_backend

    def fit(self, X, y):
        X, y = check_X_y(X, y)
        self.n_features_in_ = X.shape[1]
        self.classes_ = np.sort(np.unique(y))
        self.n_classes_ = len(self.classes_)
        if self.n_classes_ < 2:
            raise ValueError("At least two classes are required.")

        kw = dict(C=self.C, gamma=self.gamma, upsilon=self.upsilon,
                  kernel=self.kernel, degree=self.degree, coef0=self.coef0,
                  solver=self.solver, kernel_backend=self.kernel_backend)

        if self.strategy == "cs":
            self.result_ = solve_crammer_singer(X, y, **kw)
        elif self.strategy == "simmsvm":
            self.result_ = solve_simmsvm(X, y, **kw)
        elif self.strategy == "ovo":
            self.result_ = OVOClassifier(**kw).fit(X, y)
        elif self.strategy == "ovr":
            self.result_ = OVRClassifier(**kw).fit(X, y)
        else:
            raise ValueError(f"Unknown strategy: {self.strategy!r}")
        return self

    def predict(self, X):
        check_is_fitted(self, "result_")
        X = np.asarray(X, dtype=np.float64)
        if X.shape[1] != self.n_features_in_:
            raise ValueError(
                f"Expected {self.n_features_in_} features, got {X.shape[1]}."
            )
        if self.strategy == "cs":
            return _predict_cs(self.result_, X,
                                kernel_backend=self.kernel_backend)
        if self.strategy == "simmsvm":
            return predict_simmsvm(self.result_, X,
                                    kernel_backend=self.kernel_backend)
        return self.result_.predict(X)

    def decision_function(self, X):
        """Return multiclass scores (N, K). Only supported for 'cs' and 'simmsvm'."""
        check_is_fitted(self, "result_")
        if self.strategy not in ("cs", "simmsvm"):
            raise NotImplementedError(
                f"decision_function is not implemented for strategy={self.strategy!r}; "
                "use predict()."
            )
        from wrsvm.kernels import compute_kernel
        X = np.asarray(X, dtype=np.float64)
        K_new = compute_kernel(X, self.result_.X_train,
                                kernel=self.kernel, gamma=self.gamma,
                                degree=self.degree, coef0=self.coef0,
                                backend=self.kernel_backend)
        return K_new @ self.result_.theta + self.result_.b.reshape(1, -1)

File: WRSVMr/R/classifier.R
#' Fit a Weighted Relaxed SVM classifier
#'
#' Trains one of four decomposition strategies (Crammer-Singer, SimMSVM,
#' One-vs-One, One-vs-Rest) for the WRSVM formulation on multiclass data.
#' Internally calls the Python `wrsvm` package via reticulate.
#'
#' @param X Numeric matrix of shape (N, d).
#' @param y Class labels (integer or factor). Length must equal `nrow(X)`.
#' @param strategy One of `"cs"`, `"simmsvm"`, `"ovo"`, `"ovr"`. Default `"cs"`.
#' @param C Regularization parameter. Default 100.
#' @param gamma Kernel bandwidth (rbf, poly, sigmoid, laplacian). Default 0.1.
#' @param upsilon Relaxation penalty. Default 0.2.
#' @param kernel One of `"rbf"`, `"linear"`, `"poly"`, `"sigmoid"`, `"laplacian"`. Default `"rbf"`.
#' @param degree Polynomial degree (used when `kernel = "poly"`). Default 3.
#' @param coef0 Independent term for `"poly"` and `"sigmoid"` kernels. Default 0.
#' @param solver CVXPY solver name. Default `"CLARABEL"`. Use `"SCS_GPU"`
#'   for GPU acceleration when `scs` is built with CUDA.
#' @param kernel_backend One of `"numpy"` (CPU) or `"torch"` (GPU if available).
#'
#' @return A `wrsvm_model` object wrapping the Python classifier.
#' @export
#'
#' @examples
#' \dontrun{
#' data(iris)
#' X <- scale(as.matrix(iris[, 1:4]))
#' y <- as.integer(iris$Species)
#' fit <- wrsvm_fit(X, y, strategy = "cs",
#'                  C = 100, gamma = 0.1, upsilon = 0.2)
#' preds <- wrsvm_predict(fit, X)
#' mean(preds == y)
#' }
wrsvm_fit <- function(X, y,
                      strategy = "cs",
                      C = 100.0,
                      gamma = 0.1,
                      upsilon = 0.2,
                      kernel = "rbf",
                      degree = 3L,
                      coef0 = 0.0,
                      solver = "CLARABEL",
                      kernel_backend = "numpy") {
  stopifnot(is.matrix(X) || is.data.frame(X))
  X <- as.matrix(X)
  storage.mode(X) <- "double"
  y_vec <- if (is.factor(y)) as.integer(y) else y

  py <- reticulate::import("wrsvm")
  clf <- py$WRSVMClassifier(
    strategy = strategy,
    C = as.numeric(C), gamma = as.numeric(gamma), upsilon = as.numeric(upsilon),
    kernel = kernel,
    degree = as.integer(degree),
    coef0 = as.numeric(coef0),
    solver = solver, kernel_backend = kernel_backend
  )
  clf$fit(X, y_vec)

  structure(list(py_clf = clf, classes = clf$classes_,
                  strategy = strategy,
                  C = C, gamma = gamma, upsilon = upsilon,
                  kernel = kernel, degree = degree, coef0 = coef0),
            class = "wrsvm_model")
}

#' Predict class labels from a fitted WRSVM model
#'
#' @param model A `wrsvm_model` returned by [wrsvm_fit()].
#' @param X_new Numeric matrix of new samples.
#' @return Integer vector of predicted class labels.
#' @export
wrsvm_predict <- function(model, X_new) {
  stopifnot(inherits(model, "wrsvm_model"))
  X_new <- as.matrix(X_new)
  storage.mode(X_new) <- "double"
  preds <- model$py_clf$predict(X_new)
  as.vector(preds)
}

#' @export
print.wrsvm_model <- function(x, ...) {
  cat("WRSVM model\n")
  cat(sprintf("  strategy: %s\n", x$strategy))
  cat(sprintf("  kernel: %s\n", x$kernel))
  cat(sprintf("  C = %g, gamma = %g, upsilon = %g\n",
              x$C, x$gamma, x$upsilon))
  cat(sprintf("  classes: %s\n", paste(x$classes, collapse = ", ")))
  invisible(x)
}
